core/src/main/scala/kafka/server/LogOffsetMetadata.scala

```diff
@@ -58,8 +58,9 @@ case class LogOffsetMetadata(messageOffset: Long,
     this.segmentBaseOffset == that.segmentBaseOffset
   }
 
-  // check if this offset is before the given offset
-  def precedes(that: LogOffsetMetadata): Boolean = this.messageOffset < that.messageOffset
+  // check if this offset is before the given offset. We need to compare both message offset and segment base offset.
+  def precedes(that: LogOffsetMetadata): Boolean = this.messageOffset < that.messageOffset ||
+    (this.messageOffset == that.messageOffset && this.segmentBaseOffset < that.segmentBaseOffset)
```

Comment:
@ijuma That is a good point. I think we are fine here, because unknown segmentBaseOffset is -1. which works in this case. But it is clearer if we explicitly check that.

Also, LogOffsetMetadata.precedes() method is only used in tryComplete and high watermark update. The fetch logic has three conditions: 1) the fetchOffset is on an older segment than endOffset; 2) the endOffset is on an older segment than fetchOffset 3) the fetchOffset and endOffset is on the same segment.

Both 1) and 2) already checked if the it is messageOffsetOnly or not. There is a case where the fetchOffset is on an older segment, but could have been deleted when we try to send back the FetchResponse. In this case, the exception will be handled separately.